### DISCUSSION TITLE
FastAPI Background Task with DB Connection: an unexpected behaviour

### DISCUSSION THREAD

**Discussion by quamobrem** [fastapi/fastapi#12917] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends, BackgroundTasks
import aiomysql
from aiomysql import Connection, DictCursor

app = FastAPI()


""" temperatures schema
    create table `temperatures` (`id` int(11) NOT NULL AUTO_INCREMENT, `temperature` double, PRIMARY KEY(`id`)) AUTO_INCREMENT=1;
"""

async def get_connection(database: str = None):
    try:
        conn = await aiomysql.connect(
            host="0.0.0.0",
            port=3306,
            user="root",
            password="astrongpassword",
            db="db",
        )
        yield conn
    except:
        raise
    finally:
        if conn:
            conn.close()


async def get_temperatures(connection: Connection):
    async with connection.cursor(DictCursor) as cur:
        await cur.execute("SELECT * FROM temperatures;")
        return await cur.fetchall()


async def insert_temperature(temperature: float, connection: Connection):
    async with connection.cursor(DictCursor) as cur:
        await cur.execute(
            f"INSERT INTO temperatures (temperature) VALUES ({temperature});"
        )
    await connection.commit()


async def update_cache(connection: Connection):
    app.state.cache = await get_temperatures(connection=connection)


@app.get("/v1/temperatures")
async def get_cache():
    if hasattr(app.state, "cache"):
        return app.state.cache
    else:
        return []


@app.post("/v1/temperatures")
async def post_temperature(
    temperature: float,
    *,
    session: Connection = Depends(get_connection),
    background_tasks: BackgroundTasks,
):
    await insert_temperature(temperature, connection=session)

    background_tasks.add_task(update_cache, connection=session)


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(
        "minimum_reproducible_ex.api:app", host="127.0.0.1", port=8000, reload=True
    )
```

### Description

This minimally reproducible example requires a connection to a MySQL instance (8.0.0+), where user root has password "astrongpassword" and there exists a single database named `db`. 

In this example, I am leaving to a background task the onus of updating a cache after each post. 
In the MRE, this fails: `pymysql.err.InterfaceError: (0, 'Not connected')`.
I understand that this is due to the fact that after the Dependency is called during the path operation, after the response is sent, its `finally` block is executed and the connection is closed; thus when the background tasks are executed, that connection is not available. This is easily tested; if I remove the finally block, everything works (this is obviously not something that I would do).

On this point, this is really the only question I have for here: is it an antipattern, _specifically to pass a Dependecy declared via `Depends` as a task to BackgroundTasks?_ In https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#background-tasks-and-dependencies-with-yield-technical-details, I see it mentioned that it would be better if the background tasks were provisioned with their own DB connection.

Bonus:
What baffles me is a simple change I can make that I cannot explain. Please do tell me if this is better suited for the `aiomysql` forums. 
If I change the `get_connection` coroutine with:
```
async def get_connection(database: str = None):
    try:
        conn = aiomysql.connect(
            host="0.0.0.0",
            port=3306,
            user="root",
            password="astrongpassword",
            db="db",
        )
        yield await conn
    except:
        raise
    finally:
        conn.close()

```
then everything magically works. Why is `yield await` working and closing the connection after the BackgroundTasks are done?

### Operating System

Linux

### Operating System Details

Ubuntu jammy WSL

### FastAPI Version

0.111.0

### Pydantic Version

2.9.2

### Python Version

Python 3.11.0

### Additional Context

aiomysql==0.1.1

**Comment by luzzodev** [comment]
Question 1:
` if I remove the finally block, everything seems to work.`
Is not an antipattern is just a terrible practice, you are not closing the connection, so it works because connection is never closed.

Question Bonus

The difference you have in the second example is the following:
In both cases the yield value is a valid connection object but:
In the first example conn is a connection object, in the second one is just a courotine so the reason why it works is because the connection is never closed

```
async def get_connection(database: str = None):
    try:
        # conn is not a connection object but a courotine
        conn = aiomysql.connect(
            host="0.0.0.0",
            port=3306,
            user="root",
            password="astrongpassword",
            db="db",
        )
        # value here is a valid connection
        yield await conn
    except:
        raise
    finally:
        # here you have still a courotine, bonus should be awaited if is a valid connection object
        conn.close()

```

**Reply by quamobrem** [reply]
> Question 1:
> ` if I remove the finally block, everything seems to work.`
> Is not an antipattern is just a terrible practice, you are not closing the connection, so it works because connection is never closed.

I should have edited my question better, as I meant whether passing the Dependency to the BackgroundTask was an antipattern. Fixed.

**Reply by quamobrem** [reply]
> Question Bonus
> 
> The difference you have in the second example is the following: In both cases the yield value is a valid connection object but: In the first example conn is a connection object, in the second one is just a courotine so the reason why it works is because the connection is never closed

Not sure about this. `await conn.close()` crashes because `TypeError: object NoneType can't be used in 'await' expression`. 
In fact, `conn` in that case seems to be directly a subclass of the `abc.Coroutine` ABC, whose `close` method is in fact, a simple `def` method that throws a GeneritorExit.

**Comment by luzzodev** [comment, marked as answer]
I tested a bit with aiomysql i can give you more details.

What i said about your bonus question is right. As you noted close() is not aync so doesn't need to be awaited.

1) Passing argument to the background task is not an antipattern, but is a good practice to pass only arguments that are 'required' to initialize the background task ( data that is not possible to recreate in the bg task). 
Ex: if you have to run a bg task that send an email to an user X, you can set only the email as parameter and only after in the bgtask you can get the other info ( creating a connection and getting info from db and so on)

2) Bonus recap

Valid version
```
async def get_connection(database: str = None):
    try:
        conn = await aiomysql.connect(
            host="0.0.0.0",
            port=3306,
            user="root",
            password="astrongpassword",
            db="db",
        )
        yield conn
    except:
        raise
    finally:
        if conn:
            conn.close() 

```

Invalid Version: Connection not closed
```
async def get_connection(database: str = None):
    try:
    
        conn = await aiomysql.connect(
            host="0.0.0.0",
            port=3306,
            user="root",
            password="astrongpassword",
            db="db",
        )
        yield conn
    except:
        raise
```

Invalid Version: Connection not closed part 2
```
async def get_connection(database: str = None):
    try:
       # conn: aiomysql.utils._ConnectionContextManager
        conn = aiomysql.connect(
            host="0.0.0.0",
            port=3306,
            user="root",
            password="astrongpassword",
            db="db",
        )
         # passing a valid conn: aiomysql.connection.Connection
        yield await conn
    except:
        raise
    finally:
        # Here type is still : aiomysql.utils._ConnectionContextManager and you are calling close of that class.
        conn.close() 
```

Where is the trick? the real courotine is wrapped inside aiomysql.utils._ConnectionContextManager and _ConnectionContextManager is awaitable thats why you can do `await aiomysql.connect`

You can test if a connection is closed using `connection.closed` and you will see that when `it works` is because the connection have not been closed properly.

**Reply by quamobrem** [reply]
I see! 
It's actually worse in part2 because I am "loosing" the handle to the Connection object (the ConnectionContextManager coroutine has already been await so I cannot await it again and it doesn't seem like I can get access to the opened connection anymore). 
Thank you very much for the explanation.

With regards to part1, I'll just provision the background task with its own connection. I wonder if I'll have to add some sort of callback to the task to shut down the connection or using the ConnectionContextManager as a context manager will suffice.

**Reply by luzzodev** [reply]
there shouldn't be any relashionship between the endpoint and the background task.
About connection: connection shouldn't be provided or provisioned from the endpoint to the backgroundtask but should be created inside the background task. What i'm saying is that what can be retrieved or recreated in the backgroundtask should not be passed as args of bg task. (Connection, db object, data persisted shouldn't be passed as bgtask args).

 FastApi BackgroundTask are designed for now to be light and minimal without any specific constrains about when are executed.
 For bgtask that require a more fined control there are several other tools like celery or custum implementation of bg task( not fastapi dependant)

**Reply by YuriiMotov** [reply]
I would only add that it's Ok to pass something like `session_maker` through the endpoint function to the background task.

Then background task will be able to create its own session, but you will still be able to override dependency in tests

**Reply by quamobrem** [reply]
Thank you both. With both of your inputs, I have moved towards this implementation, which I am thinking is quite close to what you are suggesting, with one caveat. 

```python
class DBSessionMaker:
    def __init__(self, database: str | None = None):
        self._database = database
        self._sql_connection_obj = None

    # Making the database a getter only property makes injecting a
    # database during testing very easy
    @property
    def database(self):
        return self._database

    def _get_connection_awaitable(self) -> Awaitable[aiomysql.Connection]:
        return aiomysql.connect(
            host=...,
            port=...,
            user=...,
            password=...,
            db=self.database if self.database else ...,
        )

    async def _close_connection(self):
        try:
            logger.debug("Attempting to close MySQL connection...")
            if not self._sql_connection_obj.closed:
                await self._sql_connection_obj.ensure_closed()
        except UnboundLocalError:
            logger.critical(
                "The MySQL connection object was not created or deleted before I could close the connection!"
            )
        except Exception as exc2:
            logger.critical(f"Failed to close the MySQL connection: {exc2}")
        else:
            logger.debug("Successfully closed MySQL connection.")
            self._sql_connection_obj = None

    async def __call__(self):
        try:
            self._sql_connection_obj = await self._get_connection_awaitable()
            yield self._sql_connection_obj
        except OperationalError as connection_exc:
            logger.critical(f"Failed to yield a MySQL connection: {connection_exc}")
            raise HTTPException(
                500,
                f"Server side exception when trying to acquire a MySQL connection: {connection_exc}",
            )
        # Everything else is simply raised-through
        except Exception:
            raise
        finally:
            await self._close_connection()

    async def __aenter__(self):
        self._sql_connection_obj = await self._get_connection_awaitable()
        return self._sql_connection_obj

    async def __aexit__(self, exc_type, exc_value, traceback):
        if exc_type is not None:
            self._sql_connection_obj.close()
        else:
            await self._close_connection()


def db_provisioned_background_task(task: DBConnectionProvisionedTask, *args, **kwargs):
    async def wrapper():
        async with DBSessionMaker() as connection:
            return await task(*args, engine=connection, **kwargs)

    return wrapper
```
This works quite well: I can easily override the dependency in tests.
I went for the parametrized dependency described here approach, while also adding the `__aenter__` and `__aexit__` which allows me to use this same object as a context manager (https://fastapi.tiangolo.com/advanced/advanced-dependencies/ )
`DBConnectionProvisionedTask` is just a Protocol that defines a callable with `engine` as one of its arguments.

I think this is close to suggestions, but not quite. I can override the dependency in tests, but not in background tasks. I can still easily patch for example the `database` property in the single tests, which is fine for now. 
I understand the suggestion of @YuriiMotov as something like:

```python


class DBSessionMaker:
    # same as before plus the next method
    ...

    async def get_connection(self):
        if not self._sql_connection_obj:
            logger.debug(f"{self.database=}")
            logger.debug(
                "DBSessionMaker: don't have an existing connection open, making a new one..."
            )
            self._sql_connection_obj = await self._get_connection_awaitable()
            logger.debug("DBSessionMaker: got a new connection.")
            return self._sql_connection_obj
        else:
            logger.debug(f"{self.database=}")
            logger.debug(
                "DBSessionMaker: already have an existing connection open, returning this one."
            )
            return self._sql_connection_obj

def db_provisioned_background_task(
    task: DBConnectionProvisionedTask, session_maker: DBSessionMaker, *args, **kwargs
):
    async def wrapper():
        try:
            logger.debug("Starting background task with DB session...")
            logger.debug("Creating session for background task...")
            connection = await session_maker.get_connection()
            return await task(*args, engine=connection, **kwargs)
        finally:
            logger.debug(
                "Background task finished. Shutting down background task session..."
            )
            await session_maker._close_connection()

    return wrapper

```
this works, but the override only applies to the endpoint, and not to the background task. In guess in this case I would have to pass the class, and not an instance of it, to the endpoint? 
@YuriiMotov , would you be so kind as to provide a minimal example? I see what you mean but I can't quite replicate it. 
Thank you both!
